### PR TITLE
UI: Current Image Display - Smooth Preview

### PR DESCRIPTION
**Pull request by blessedcoolant**
## What type of PR is this? (check all applicable)

- [x] Feature

## Have you discussed this change with the InvokeAI team?
- [x] No

      
## Description

- Changes image saving to CV2 instead of PIL. Shows better performance for me. Can be reverted if this is not the general case.
- Hacky effort to make the Current Image Display smoother with transitions and animation. Looks neat imo.

https://github.com/invoke-ai/InvokeAI/assets/54517381/4b9191ef-5696-4ee8-ab9f-215ff1d037f1

### CODE REVIEW

**Comment by hipsterusername** [comment]
Looks great! Helps with perception a lot

**Comment by hipsterusername** [comment]
One thing to test is if there is an issue on the non-happy path (probably vae), whether it intelligently bails on the “smooth transition” - As well as how it handles something like hi-res workflows

File: invokeai/app/services/image_files/image_files_disk.py
Comment (psychedelicious) [inline invokeai/app/services/image_files/image_files_disk.py:1]: In my testing, PIL was faster than OpenCV at across the board. It also produced smaller files. I'm not sure we should change this

**Review by psychedelicious — CHANGES_REQUESTED** [review]
I can test out tomorrow but there are edge cases we need to think about. Off the top of my head:
- Hi res fix?
- When you have a graph with multiple denoise nodes?
- Interaction with the progress toggle?

**Comment by lstein** [comment]
Does this fix the awkward transition between the last denoise step preview and the final VAE decode? I have never liked the fact that the previous image is redisplayed during this transition.

File: invokeai/app/services/image_files/image_files_disk.py
Comment (psychedelicious) [reply invokeai/app/services/image_files/image_files_disk.py:1]: Here's my test: https://gist.github.com/psychedelicious/78f0f71cdde2e4efafa2fac4e4c2c953

It uses random pixels, so it's not a real-world test. It should be adapted to work on real image data. This is probably only showing worse-case performance.

**Comment by blessedcoolant** [comment]
> Does this fix the awkward transition between the last denoise step preview and the final VAE decode? I have never liked the fact that the previous gallery image is redisplayed during this transition.

Yes it does.

File: invokeai/app/services/image_files/image_files_disk.py
Comment (blessedcoolant) [reply invokeai/app/services/image_files/image_files_disk.py:1]: Did not check hi-res fix. I will do it now.

When there are two Denoise Latents, the result of the first one is not shown. Only the one at the end of the graph.

I think the answer for this is not adjusting this but rather actually creating a proper preview node for the graph whose exclusive job is to display the image. There is an absolute necessity for it in terms of the graph itself.

As for PIL or OpenCV, I think it'll be fine either way. If you feel we want to stick to PIL, then that is fine too. Maybe we can let others test out and see what is more performant for them and then decide based on that.

Comment (blessedcoolant) [reply invokeai/app/services/image_files/image_files_disk.py:1]: There was some odd cut off happening with hi-res fix but not sure what was causing it. Anyway, fixed it with a proper solution anyway. All good on that end I think.

Comment (blessedcoolant) [reply invokeai/app/services/image_files/image_files_disk.py:1]: Interaction with progress toggle is smooth.

Comment (psychedelicious) [reply invokeai/app/services/image_files/image_files_disk.py:1]: There's some jank when the selected image is larger than the currently-generating image 

Uploading Screen Recording 2023-11-30 at 4.17.18 pm.mov…

Comment (psychedelicious) [reply invokeai/app/services/image_files/image_files_disk.py:1]: Ok so the prev image still flashes through briefly

https://github.com/invoke-ai/InvokeAI/assets/4822129/2d2a8cc7-dac5-4644-b88f-ab601ca2286d

File: invokeai/frontend/web/src/features/system/store/systemSlice.ts
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/system/store/systemSlice.ts:1]: I *think* the changes here are fine. At some point in the past the nulling was necessary but seems fine now. Cool!

File: invokeai/app/services/image_files/image_files_disk.py
Comment (psychedelicious) [reply invokeai/app/services/image_files/image_files_disk.py:1]: After some solid testing, I think we could switch to cv2 & set the compress level to 1. But there are two issues.

Testing: https://discord.com/channels/1020123559063990373/1179252598520418334/1179646194910240820

1) We now have to convert the PIL image to np array, then swap the channel. I assume this takes virtually no time but still it's an extra step.

2) Other issue - the changes in this PR do not write metadata or workflow to the image file.

**Comment by psychedelicious** [comment]
There's still a flash of the previous image for me. It's brief but kinda disorienting, especially when the images are of different sizes

https://github.com/invoke-ai/InvokeAI/assets/4822129/81f6066f-92a3-480f-88a4-6cc8a5717ae3

File: invokeai/app/services/image_files/image_files_disk.py
Comment (blessedcoolant) [reply invokeai/app/services/image_files/image_files_disk.py:1]: Ye I don't know how we can dodge that. Some really fancy swoop in and out animation maybe. But could be overkill.

Comment (blessedcoolant) [reply invokeai/app/services/image_files/image_files_disk.py:1]: Initially I was trying to dodge the np array conversion but I realize there's a ton of nodes that directly work with PIL images and I feel changing all of that is a much bigger hassle currently.

Comment (blessedcoolant) [reply invokeai/app/services/image_files/image_files_disk.py:1]: I'll take a look at the metadata / workflow part. Not sure which part affected it. Don't think I touched anything regards to that.

**Comment by psychedelicious** [comment]
This would close #3987

**Review by lstein — COMMENTED** [review]
Code looks clean to me. Tested on two browsers and the iPad and the transition is very smooth, even when changing image sizes.

However, I can confirm that the generation metadata is not being written into the png file.

**Review by lstein — APPROVED** [review]
I made a one-line commit that restores metadata writing into the image file. It does not seem that opencv2 supports writing PNG text blocks (of any flavor), so the change reopens the file as a PIL::Image and immediately saves it again with the metadata.

File: invokeai/frontend/web/src/features/system/store/systemSlice.ts
Comment (psychedelicious) [reply invokeai/frontend/web/src/features/system/store/systemSlice.ts:1]: Actually, this doesn't work. It waits for the whole graph to be completed before showing the "final" output image.

Imagine you have a denoise node then other nodes for some amount of time until the next image-outputting node. I've simulated this by adding a `sleep(10)` to the latents primitive and putting it between the denoise and decode:
image

The result is that when denoise finishes, the progress image stays as the last progress image until the graph is completed.

What if there were other image-outputting nodes between those two? Or elsewhere in the graph?

**Comment by blessedcoolant** [comment]
I think this is in a good place. If @psychedelicious is good with the changes, we can go ahead and merge. @hipsterusername .. The other changes in regards to how to handle previews in general could go with another PR.

**Comment by hipsterusername** [comment]
I'll test and confirm that it doesn't have the flicker! If so, I'll approve and merge

File: invokeai/app/services/image_files/image_files_disk.py [hunk 1]
```diff
@@ -77,6 +77,9 @@ def save(
                 compress_level=self.__invoker.services.configuration.png_compress_level,
             )
 
+            # write png info using PIL
+            Image.open(image_path.as_posix()).save(image_path.as_posix(), pnginfo=pnginfo)
+
```
Comment (psychedelicious) [inline invokeai/app/services/image_files/image_files_disk.py:82]: I think this is a leftover from testing with CV2

**Comment by psychedelicious** [comment]
The weird flash related to the animation and aspect ratios is, IMO, reason enough to not include it. Looks very janky

**Comment by blessedcoolant** [comment]
All good. We can wait for a better solution. Probably something that'll resolve itself when we isolate the previews.

**Comment by hipsterusername** [comment]
@psychedelicious - If you can think of a short-term resolution alternative to the jank, would be a good thing.
